File: LC-Python-Solution/160. Intersection of Two Linked Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:

        def getLength(head):
            current = head
            length = 0
            while current:
                length+=1
                current = current.next
            return length

        lenA = getLength(headA)
        lenB = getLength(headB)

        diff = abs(lenA-lenB)

        if(lenA>lenB):
            for i in range(diff):
                headA = headA.next
        else:
            for i in range(diff):
                headB = headB.next
        
        while headA and headB:
            if headA == headB:
                return headA
            headA = headA.next
            headB = headB.next
        
        return None
            




        

File: LC_JS-SOLUTIONS/160. Intersection of Two Linked Lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var getIntersectionNode = function (headA, headB) {
  function getLength(head) {
    let length = 0;
    let current = head;
    while (current) {
      length += 1;
      current = current.next;
    }
    return length;
  }

  let lenA = getLength(headA);
  let lenB = getLength(headB);

  let diff = Math.abs(lenA - lenB);

  if (lenA > lenB) {
    for (i = 0; i < diff; i++) {
      headA = headA.next;
    }
  } else if (lenB > lenA) {
    for (i = 0; i < diff; i++) {
      headB = headB.next;
    }
  }

  while (headA && headB) {
    if (headA === headB) {
      return headA;
    }
    headA = headA.next;
    headB = headB.next;
  }
  return null;
};
